iceoryx_utils/include/iceoryx_utils/lockfree_queue/lockfree_queue.hpp

```diff
@@ -0,0 +1,123 @@
+// Copyright (c) 2019 by Robert Bosch GmbH. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include "iceoryx_utils/cxx/optional.hpp"
+#include "iceoryx_utils/internal/lockfree_queue/buffer.hpp"
+#include "iceoryx_utils/internal/lockfree_queue/index_queue.hpp"
+
+#include <atomic>
+
+
+namespace iox
+{
+/// @todo finalize interface, configuration of actual capacity at runtime (change capacity feature)
+
+/// @brief implements a lock free queue (i.e. container with FIFO order) of elements of type T
+/// with Capacity
+template <typename ElementType, uint64_t Capacity>
+class LockFreeQueue
+{
+  public:
+    /// @brief creates and initalizes an empty LockFreeQueue
+    LockFreeQueue() noexcept;
+
+    /// @todo: a thread-safe and lockfree implementation of copy seems impossible
+    /// but unsafe copying (i.e. where synchronization is up to the user) would be possible
+    /// can be implemented when it is needed
+    LockFreeQueue(const LockFreeQueue&) = delete;
+    LockFreeQueue(LockFreeQueue&&) = delete;
+    LockFreeQueue& operator=(const LockFreeQueue&) = delete;
+    LockFreeQueue& operator=(LockFreeQueue&&) = delete;
+
+    /// @brief returns the capacity of the queue
+    /// threadsafe, lockfree
+    constexpr uint64_t capacity() noexcept;

```

Comment:
I think you should make this method const

Reply:
Yes